### PR TITLE
[wasm] Endian fix for Webcil

### LINKED ISSUE #91893: The "ConvertDllsToWebCil" task failed:  System.NotImplementedException on s390x
### Description

With a .NET 8 Preview 7 SDK  on s390x (IBM Z), running `dotnet new blazorwasm && dotnet build` fails:

```
/usr/lib64/dotnet/sdk/8.0.100-preview.7.23376.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm_template/blazorwasm_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: The "ConvertDllsToWebCil" task failed unexpectedly. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: System.NotImplementedException: The method or operation is not implemented. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteStructure[WebcilHeader](Stream s, WebcilHeader structure) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteHeader(Stream s, WebcilHeader header) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteConversionTo(Stream outputStream, FileStream inputStream, PEFileInfo peInfo, WCFileInfo wcInfo) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.WebAssembly.Build.Tasks.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.Sdk.WebAssembly.ConvertDllsToWebCil.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
```

As does `dotnet new blzaorwasm-empty && dotnet build`:

```
/usr/lib64/dotnet/sdk/8.0.100-preview.7.23376.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(314,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: The "ConvertDllsToWebCil" task failed unexpectedly. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018: System.NotImplementedException: The method or operation is not implemented. [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteStructure[WebcilHeader](Stream s, WebcilHeader structure) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteHeader(Stream s, WebcilHeader header) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteConversionTo(Stream outputStream, FileStream inputStream, PEFileInfo peInfo, WCFileInfo wcInfo) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.WebAssembly.Build.Tasks.WebcilConverter.ConvertToWebcil() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.NET.Sdk.WebAssembly.ConvertDllsToWebCil.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
/root/.nuget/packages/microsoft.net.sdk.webassembly.pack/8.0.0-preview.7.23375.6/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets(247,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/var/tmp/tmp.ZoIlFRbKGK/blazorwasm-empty_template/blazorwasm-empty_template.csproj]
```

### Reproduction Steps

`dotnet new blazorwasm && dotnet build` on IBM Z

### Expected behavior

The `dotnet build` works, or the `blazorwasm` entry doesn't show up as a valid template item.

### Actual behavior

Stack trace.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

It's probably from here, which happens when the architecture is big-endian only: https://github.com/dotnet/runtime/blob/9d08b24d743d0c57203a55c3d0c6dd1bc472e57e/src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs#L217

### PR DESCRIPTION
**Pull request by saitama951**
'dotnet new blazorwasm' command failed on s390x and was throwing a not implemented exception

The issue was with with the WebCil writer and reader, specific endianness conversions relating to the webcil payload were not implemented for big endian machines.

We considered fixing the generic implementation, but there were only two structures in use: WebcilHeader and WebcilSectionHeader, so it was easier to handle them explicitly.

fixes #91893 
cc: @omajid @iii-i @uweigand

### CODE REVIEW

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs [hunk 1]
```diff
@@ -181,33 +182,61 @@ private static void WriteHeader(Stream s, WebcilHeader header)
...
+    private static void WriteSectionHeader(Stream s, WebcilSectionHeader sectionHeader)
+    {
+        if (!BitConverter.IsLittleEndian)
+        {
+            sectionHeader = new WebcilSectionHeader
+            (
+                virtualSize: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(sectionHeader.VirtualSize)),
+                virtualAddress: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(sectionHeader.VirtualAddress)),
+                sizeOfRawData: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(sectionHeader.SizeOfRawData)),
+                pointerToRawData: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(sectionHeader.PointerToRawData))
```
Comment (teo-tsirpanis) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs:201]:
```suggestion
                virtualSize: BinaryPrimitives.ReverseEndianess(sectionHeader.VirtualSize),
                virtualAddress: BinaryPrimitives.ReverseEndianess(sectionHeader.VirtualAddress),
                sizeOfRawData: BinaryPrimitives.ReverseEndianess(sectionHeader.SizeOfRawData),
                pointerToRawData: BinaryPrimitives.ReverseEndianess(sectionHeader.PointerToRawData)
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs [hunk 2]
```diff
@@ -181,33 +182,61 @@ private static void WriteHeader(Stream s, WebcilHeader header)
...
+    {
+        if (!BitConverter.IsLittleEndian)
+        {
+            webcilHeader.version_major = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes( webcilHeader.version_major));
+            webcilHeader.version_minor = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes( webcilHeader.version_minor));
+            webcilHeader.coff_sections = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes(webcilHeader.coff_sections));
+            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(webcilHeader.pe_cli_header_rva));
+            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(webcilHeader.pe_cli_header_size));
+            webcilHeader.pe_debug_rva = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(webcilHeader.pe_debug_rva));
+            webcilHeader.pe_debug_size = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(webcilHeader.pe_debug_size));
```
Comment (teo-tsirpanis) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs:236]:
```suggestion
            webcilHeader.version_major = BinaryPrimitives.ReverseEndianess(webcilHeader.version_major);
            webcilHeader.version_minor = BinaryPrimitives.ReverseEndianess(webcilHeader.version_minor);
            webcilHeader.coff_sections = BinaryPrimitives.ReverseEndianess(webcilHeader.coff_sections);
            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReverseEndianess(webcilHeader.pe_cli_header_rva);
            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReverseEndianess(webcilHeader.pe_cli_header_size);
            webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianess(webcilHeader.pe_debug_rva);
            webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianess(webcilHeader.pe_debug_size);
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs [hunk 3]
```diff
@@ -63,14 +63,20 @@ private unsafe bool ReadHeader()
...
         }
+        if (!BitConverter.IsLittleEndian)
+        {
+            header.version_major = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes(header.version_major));
+            header.version_minor = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes(header.version_minor));
+            header.coff_sections = BinaryPrimitives.ReadUInt16LittleEndian(BitConverter.GetBytes(header.coff_sections));
+            header.pe_cli_header_rva = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(header.pe_cli_header_rva));
+            header.pe_cli_header_size = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(header.pe_cli_header_size));
+            header.pe_debug_rva = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(header.pe_debug_rva));
+            header.pe_debug_rva = BinaryPrimitives.ReadUInt32LittleEndian(BitConverter.GetBytes(header.pe_debug_size));
```
Comment (teo-tsirpanis) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs:78]:
```suggestion
            header.version_major = BinaryPrimitives.ReverseEndianess(header.version_major);
            header.version_minor = BinaryPrimitives.ReverseEndianess(header.version_minor);
            header.coff_sections = BinaryPrimitives.ReverseEndianess(header.coff_sections);
            header.pe_cli_header_rva = BinaryPrimitives.ReverseEndianess(header.pe_cli_header_rva);
            header.pe_cli_header_size = BinaryPrimitives.ReverseEndianess(header.pe_cli_header_size);
            header.pe_debug_rva = BinaryPrimitives.ReverseEndianess(header.pe_debug_rva);
            header.pe_debug_rva = BinaryPrimitives.ReverseEndianess(header.pe_debug_size);
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs [hunk 4]
```diff
@@ -357,8 +364,24 @@ private unsafe ImmutableArray<WebcilSectionHeader> ReadSections()
...
+            if (!BitConverter.IsLittleEndian)
+            {
+                sections.Add
+                (
+                    new WebcilSectionHeader
+                    (
+                        virtualSize: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(secheader.VirtualSize)),
+                        virtualAddress: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(secheader.VirtualAddress)),
+                        sizeOfRawData: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(secheader.SizeOfRawData)),
+                        pointerToRawData: BinaryPrimitives.ReadInt32LittleEndian(BitConverter.GetBytes(secheader.PointerToRawData))
```
Comment (teo-tsirpanis) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilReader.cs:378]:
```suggestion
                        virtualSize: BinaryPrimitives.ReverseEndianess(secheader.VirtualSize),
                        virtualAddress: BinaryPrimitives.ReverseEndianess(secheader.VirtualAddress),
                        sizeOfRawData: BinaryPrimitives.ReverseEndianess(secheader.SizeOfRawData),
                        pointerToRawData: BinaryPrimitives.ReverseEndianess(secheader.PointerToRawData)
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs [hunk 5]
```diff
@@ -181,33 +182,61 @@ private static void WriteHeader(Stream s, WebcilHeader header)
...
-        // FIXME: fixup endianness
-        if (!BitConverter.IsLittleEndian)
-            throw new NotImplementedException();
         foreach (var sectionHeader in sectionsHeaders)
         {
             WriteSectionHeader(s, sectionHeader);
         }
     }
 
+#if NETCOREAPP2_1_OR_GREATER
```
Comment (teo-tsirpanis) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs:191]: Thanks. Can you try removing the `#if` if it compiles? This function might be available on all frameworks (unless I'm missing something).

Comment (saitama951) [reply src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs]: Thanks for pointing that out. I have done the same for the WriteStructure as well and hence removing the checks for net472 as well.

**Review by lambdageek — APPROVED** [review]
thanks!


**Update** Please add the `PackageReference` to the csproj (see https://github.com/dotnet/runtime/pull/92274#discussion_r1331650413)

**Comment by saitama951** [comment]
> thanks!
> 
> **Update** Please add the `PackageReference` to the csproj (see [#92274 (comment)](https://github.com/dotnet/runtime/pull/92274#discussion_r1331650413))

@lambdageek I have done the suggested changes, Thank you for the input.

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
   <ItemGroup>
     <!-- we need to keep the version of System.Reflection.Metadata in sync with dotnet/msbuild and dotnet/sdk -->
+    <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
```
Comment (radical) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/Microsoft.NET.WebAssembly.Webcil.csproj:19]:
```suggestion
    <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs [hunk 7]
```diff
@@ -192,29 +190,49 @@ private static void WriteSectionHeaders(Stream s, ImmutableArray<WebcilSectionHe
...
-        // FIXME: fixup endianness
-        if (!BitConverter.IsLittleEndian)
-            throw new NotImplementedException();
         unsafe
         {
             byte* p = (byte*)&structure;
             s.Write(new ReadOnlySpan<byte>(p, sizeof(T)));
         }
     }
+
```
Comment (radical) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs:231]: nit: seems unnecessary

Comment (saitama951) [reply src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs]: done

**Comment by saitama951** [comment]
@lambdageek @radical @teo-tsirpanis can this be merged?

**Comment by radical** [comment]
WasmBuildTests has relevant failures, [like](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-92274-merge-d7b13bcfb56246e686/Workloads-Wasm.Build.NativeRebuild.Tests.FlagsChangeRebuildTests/1/console.10297b39.log?helixlogtype=result):

```
      [] Stack overflow.
      [] Repeat 74781 times:
      [] --------------------------------
      []    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteStructure(System.IO.Stream, Microsoft.NET.WebAssembly.Webcil.WebcilHeader)
      [] --------------------------------
      []    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteHeader(System.IO.Stream, Microsoft.NET.WebAssembly.Webcil.WebcilHeader)
      []    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.WriteConversionTo(System.IO.Stream, System.IO.FileStream, PEFileInfo, WCFileInfo)
      []    at Microsoft.NET.WebAssembly.Webcil.WebcilConverter.ConvertToWebcil()
      []    at Microsoft.WebAssembly.Build.Tasks.WebcilConverter.ConvertToWebcil()
      []    at Microsoft.WebAssembly.Build.Tasks.WasmAppBuilder.ExecuteInternal()
      []    at Microsoft.WebAssembly.Build.Tasks.WasmAppBuilderBaseTask.Execute()
      []    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
      []    at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteInstantiatedTask>d__25.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TaskBuilder+<ExecuteInstantiatedTask>d__25, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ExecuteInstantiatedTask>d__25 ByRef)
      []    at Microsoft.Build.BackEnd.TaskBuilder+<InitializeAndExecuteTask>d__23.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TaskBuilder+<InitializeAndExecuteTask>d__23, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<InitializeAndExecuteTask>d__23 ByRef)
      []    at Microsoft.Build.BackEnd.TaskBuilder.InitializeAndExecuteTask(Microsoft.Build.BackEnd.Logging.TaskLoggingContext, Microsoft.Build.BackEnd.ItemBucket, System.Collections.Generic.IDictionary`2<System.String,System.String>, Microsoft.Build.BackEnd.TaskHost, Microsoft.Build.BackEnd.TaskExecutionMode)
      []    at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteBucket>d__19.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TaskBuilder+<ExecuteBucket>d__19, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ExecuteBucket>d__19 ByRef)
      []    at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteTask>d__18.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TaskBuilder+<ExecuteTask>d__18, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ExecuteTask>d__18 ByRef)
      []    at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteTask>d__13.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TaskBuilder+<ExecuteTask>d__13, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ExecuteTask>d__13 ByRef)
      []    at Microsoft.Build.BackEnd.TargetEntry+<ProcessBucket>d__51.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TargetEntry+<ProcessBucket>d__51, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ProcessBucket>d__51 ByRef)
      []    at Microsoft.Build.BackEnd.TargetEntry+<ExecuteTarget>d__44.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TargetEntry+<ExecuteTarget>d__44, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ExecuteTarget>d__44 ByRef)
      []    at Microsoft.Build.BackEnd.TargetEntry.ExecuteTarget(Microsoft.Build.BackEnd.ITaskBuilder, Microsoft.Build.BackEnd.BuildRequestEntry, Microsoft.Build.BackEnd.Logging.ProjectLoggingContext, System.Threading.CancellationToken)
      []    at Microsoft.Build.BackEnd.TargetBuilder+<ProcessTargetStack>d__23.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TargetBuilder+<ProcessTargetStack>d__23, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ProcessTargetStack>d__23 ByRef)
      []    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.Build.BackEnd.TargetBuilder+<ProcessTargetStack>d__23, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<ProcessTargetStack>d__23 ByRef)
      []    at Microsoft.Build.BackEnd.TargetBuilder.ProcessTargetStack(Microsoft.Build.BackEnd.ITaskBuilder)
      []    at Microsoft.Build.BackEnd.TargetBuilder+<BuildTargets>d__10.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.TargetBuilder+<BuildTargets>d__10, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildTargets>d__10 ByRef)
      []    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Microsoft.Build.BackEnd.TargetBuilder+<BuildTargets>d__10, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildTargets>d__10 ByRef)
      []    at Microsoft.Build.BackEnd.TargetBuilder.BuildTargets(Microsoft.Build.BackEnd.Logging.ProjectLoggingContext, Microsoft.Build.BackEnd.BuildRequestEntry, Microsoft.Build.BackEnd.IRequestBuilderCallback, System.String[], Microsoft.Build.BackEnd.Lookup, System.Threading.CancellationToken)
      []    at Microsoft.Build.BackEnd.RequestBuilder+<BuildProject>d__68.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.RequestBuilder+<BuildProject>d__68, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildProject>d__68 ByRef)
      []    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Start[[Microsoft.Build.BackEnd.RequestBuilder+<BuildProject>d__68, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildProject>d__68 ByRef)
      []    at Microsoft.Build.BackEnd.RequestBuilder.BuildProject()
      []    at Microsoft.Build.BackEnd.RequestBuilder+<BuildAndReport>d__59.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.RequestBuilder+<BuildAndReport>d__59, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildAndReport>d__59 ByRef)
      []    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.Build.BackEnd.RequestBuilder+<BuildAndReport>d__59, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<BuildAndReport>d__59 ByRef)
      []    at Microsoft.Build.BackEnd.RequestBuilder.BuildAndReport()
      []    at Microsoft.Build.BackEnd.RequestBuilder+<RequestThreadProc>d__58.MoveNext()
      []    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[Microsoft.Build.BackEnd.RequestBuilder+<RequestThreadProc>d__58, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<RequestThreadProc>d__58 ByRef)
      []    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.Build.BackEnd.RequestBuilder+<RequestThreadProc>d__58, Microsoft.Build, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]](<RequestThreadProc>d__58 ByRef)
      []    at Microsoft.Build.BackEnd.RequestBuilder.RequestThreadProc(Boolean)
      []    at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
      []    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
      []    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
      []    at System.Threading.Tasks.Task.ExecuteEntry()
      []    at Microsoft.Build.BackEnd.RequestBuilder+DedicatedThreadsTaskScheduler.<InjectThread>b__6_0()
      []    at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
```

File: src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs [hunk 8]
```diff
@@ -192,16 +190,38 @@ private static void WriteSectionHeaders(Stream s, ImmutableArray<WebcilSectionHe
...
+        {
+            webcilHeader.version_major = BinaryPrimitives.ReverseEndianness(webcilHeader.version_major);
+            webcilHeader.version_minor = BinaryPrimitives.ReverseEndianness(webcilHeader.version_minor);
+            webcilHeader.coff_sections = BinaryPrimitives.ReverseEndianness(webcilHeader.coff_sections);
+            webcilHeader.pe_cli_header_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_rva);
+            webcilHeader.pe_cli_header_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_cli_header_size);
+            webcilHeader.pe_debug_rva = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_rva);
+            webcilHeader.pe_debug_size = BinaryPrimitives.ReverseEndianness(webcilHeader.pe_debug_size);
+        }
+        WriteStructure(s, webcilHeader);
```
Comment (lambdageek) [inline src/tasks/Microsoft.NET.WebAssembly.Webcil/WebcilConverter.cs:218]:
```suggestion
        } else
        WriteStructure(s, webcilHeader);
```

**Review by lambdageek — COMMENTED** [review]
oops

**Comment by lambdageek** [comment]
Moved the endianess fixup for `WebcilHeader` to `WriteHeader` instead of an overload of `WriteStructure` /cc @saitama951

**Comment by uweigand** [comment]
This is failing in .NET 8 - can this be merged there?

**Pull request merged**